### PR TITLE
use SEP-29 to check for memo requirement

### PR DESCRIPTION
**Pull request by piyalbasu**
Screen Shot 2021-03-04 at 4 00 48 PM
Screen Shot 2021-03-04 at 4 00 40 PM

### CODE REVIEW

File: extension/src/background/messageListener/freighterApiMessageListener.ts [hunk 1]
```diff
@@ -92,21 +93,31 @@ export const freighterApiMessageListener = (
 
     const { _operations } = transaction;
 
-    const flaggedKeys = _operations.reduce(
-      (arr: Array<{}>, operation: { destination: string }) => {
-        const listing = accountData.find(
-          ({ address }: { address: string }) =>
-            address === operation.destination,
-        );
+    const flaggedKeys: FlaggedKeys = {};
```
Comment (piyalbasu) [inline extension/src/background/messageListener/freighterApiMessageListener.ts:96]: changing `FlaggedKeys` to an object prevents multiple entries that pertain to the same public key. Now, we will just update a public key's entry if we find more info about in on an iteration

File: extension/src/popup/components/signTransaction/Operations.tsx [hunk 2]
```diff
@@ -171,6 +166,49 @@ const DestinationWarning = ({
...
+            alt="exclamation icon"
+            icon={IconExcalamtion}
+          >
+            Memo required
+          </IconWithLabel>
+        </OpertionValueExtraEl>
+      }
+    />
+  ) : null;
+};
```
Comment (piyalbasu) [inline extension/src/popup/components/signTransaction/Operations.tsx:190]: adding a warning to each Operation that requires a memo

File: extension/src/popup/components/warningMessages/FlaggedWarningMessage.tsx [hunk 3]
```diff
@@ -56,9 +56,9 @@ const MemoWarningMessage = ({ isMemoRequired }: { isMemoRequired: boolean }) =>
       subheader="Memo is required"
     >
       <p>
-        The destination account (Kraken) requires the use of the memo field
-        which is not present in the transaction you're about to sign. Freighter
-        automatically disabled the option to sign this transaction.
+        A destination account requires the use of the memo field which is not
+        present in the transaction you’re about to sign. Freighter automatically
+        disabled the option to sign this transaction.
```
Comment (piyalbasu) [inline extension/src/popup/components/warningMessages/FlaggedWarningMessage.tsx:61]: Updating copy to reflect that one (of possibly several operations) threw an error

File: extension/src/popup/views/SignTransaction.tsx [hunk 4]
```diff
@@ -138,27 +139,35 @@ export const SignTransaction = () => {
...
-  const isMemoRequired = flaggedKeys.some(({ tags }) =>
-    tags.includes("memo-required"),
+  const isMalicious = flaggedKeyValues.some(({ tags }) =>
+    tags.includes(TRANSACTION_WARNING.malicious),
   );
+  const isMemoRequired = flaggedKeyValues.some(({ tags }) =>
+    tags.includes(TRANSACTION_WARNING.memoRequired),
+  );
+
+  const isSubmitDisabled = isMemoRequired || isMalicious;
```
Comment (piyalbasu) [inline extension/src/popup/views/SignTransaction.tsx:154]: memo-required should prevent submission

File: extension/src/popup/components/signTransaction/Operations.tsx [hunk 5]
```diff
@@ -1,8 +1,9 @@
 import React from "react";
 import styled from "styled-components";
 
-import { OPERATION_TYPES } from "constants/operationTypes";
+import { OPERATION_TYPES } from "constants/transaction";
 import { COLOR_PALETTE, FONT_WEIGHT } from "popup/constants/styles";
+import { TRANSACTION_WARNING } from "constants/transaction";
```
Comment (quietbits) [inline extension/src/popup/components/signTransaction/Operations.tsx:6]: Is this the same source as on line #4?

**Review by quietbits — APPROVED** [review]

File: extension/src/popup/components/signTransaction/Operations.tsx (hunk 5 above)
Comment (piyalbasu) [reply extension/src/popup/components/signTransaction/Operations.tsx:6]: Oh yeah - good catch! Weird that the linter didn't pick that up 🤔

**Pull request merged**